### PR TITLE
feat(backend): dynamic scrap target (over config file)

### PR DESCRIPTION
**Pull request by ii64**
* dynamic scrap target
* changed global hashmap to struct
* github optional repo description
* graceful exit for config write-back

### CODE REVIEW

File: backend/src/config.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    fn test_decode_config() {
+
+    }
+    fn test_encode_config() {
+
+    }
```
Comment (aldy505) [inline backend/src/config.rs:94]: ini WIP?

File: backend/src/scraper.rs [hunk 2]
```diff
@@ -28,90 +32,155 @@ pub struct RepositoryCollection {
...
-impl Display for ScrapError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self)
-    }
-}
+// impl Display for ScrapError {
+//     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+//         write!(f, "{}", self)
+//     }
+// }
```
Comment (aldy505) [inline backend/src/scraper.rs:45]: delete aja, jangan commented out

File: backend/src/main.rs [hunk 3]
```diff
@@ -24,14 +33,44 @@ async fn main() {
     }
 }
 
+pub struct RunContextInner {
+}
+impl RunContextInner {
+}
```
Comment (aldy505) [inline backend/src/main.rs:39]: ini kenapa kosong?

File: backend/src/config.rs (hunk 1 above)
Comment (ii64) [reply backend/src/config.rs:94]: Mmm yeah

File: backend/src/main.rs (hunk 3 above)
Comment (ii64) [reply backend/src/main.rs:39]: Placeholder buat lifetime RunContext, also buat `mpsc` (still in doubt)

File: .github/workflows/PR.yml [hunk 4]
```diff
@@ -8,7 +8,7 @@ jobs:
     name: Backend CI
     runs-on: ubuntu-latest
     container: rust:1.63.0-bullseye
-    timeout-minutes: 10
+    timeout-minutes: 11
```
[the remark this answers is not in the document]
Comment (aldy505) [reply .github/workflows/PR.yml:11]: sekalian aja jadi 20 menit kalo gitu mah wkwkwkwk

[the remark this answers is not in the document]
Comment (ii64) [reply .github/workflows/PR.yml:11]: Woke

**Review by aldy505 — CHANGES_REQUESTED** [review]

File: backend/src/main.rs [hunk 5]
```diff
@@ -4,58 +4,101 @@ use actix_web::{App, HttpServer, Result};
...
 
 use crate::handlers::*;
-use crate::defaults::*;
 
 #[tokio::main]
 async fn main() {
+    if cfg!(debug_assertions) {
+        color_backtrace::install()
+    }
+    env_logger::init();
```
[the remark this answers is not in the document]
Comment (ii64) [reply backend/src/main.rs:33]: Adjust log level at run-time, `RUST_LOG=hacktoberfest=<level>`

**Comment by aldy505** [comment]
ah i see, boleh boleh

File: backend/src/github.rs [hunk 6]
```diff
@@ -222,9 +225,10 @@ impl Github {
 
     /// list_issues
     ///
-    pub async fn list_issues(&self, repo: String) -> Result<Vec<Issue>, GithubError> {
+    pub async fn list_issues(&self, user: String, repo: String) -> Result<Vec<Issue>, GithubError> {
+        let uencoded_user = urlencoding::encode(&user[..]);
```
[the remark this answers is not in the document]
Comment (ii64) [reply backend/src/github.rs:229]: Alright then xD

File: backend/src/github.rs [hunk 7]
```diff
@@ -209,11 +209,14 @@ impl Github {
...
-    pub async fn list_repository(&self) -> Result<Vec<Repository>, GithubError> {
+    pub async fn list_repository(&self, user: String) -> Result<Vec<Repository>, GithubError> {
+        let uencoded_user = urlencoding::encode(&user[..]);
+        let u = format!("https://api.github.com/users/{uencoded_user}/repos");
         let response: Response = self
             .client
-            .get("https://api.github.com/users/teknologi-umum/repos")
-            .query(&[("type", "public"), ("sort", "updated"), ("per_page", "100")])
+            .get(u)
+            // Figure out what to do with `per_page`? hard coded for now..
```
[the remark this answers is not in the document]
Comment (ii64) [reply backend/src/github.rs:218]: Oke default 100, adjustable with CLI / env now

**Review by aldy505 — APPROVED** [review]

**Comment by aldy505** [comment]
@ii64 masih WIP?

**Comment by ii64** [comment]
> @ii64 masih WIP?

Merge time

File: backend/src/config.rs [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        }
+    }
+    
+    pub fn ignore(mut self) -> Self {
+        self.ignore = true;
+        self
+    }
+    pub fn target_links(&self) -> Vec<String> {
+        self.repo_names.as_ref()
+            .unwrap_or(&Vec::with_capacity(8))
```
Comment (elianiva) [inline backend/src/config.rs:47]: kenapa 8?

**Review by elianiva — APPROVED** [review]
one question, sisanya udah oke

File: backend/src/config.rs (hunk 8 above)
Comment (ii64) [reply backend/src/config.rs:47]: Oh yea, no need to have extra allocation here

**Pull request merged**